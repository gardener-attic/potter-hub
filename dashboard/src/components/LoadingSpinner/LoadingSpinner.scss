$spinner-color: #00437b;

.spinner {
  height: 40px;
  margin: 100px auto;
  position: relative;
  width: 40px;

  &__bounce1,
  &__bounce2 {
    animation: spinner-bounce 2s infinite ease-in-out;
    background-color: $spinner-color;
    border-radius: 50%;
    height: 100%;
    left: 0;
    opacity: 0.6;
    position: absolute;
    top: 0;
    width: 100%;
  }

  &__bounce2 {
    animation-delay: -1s;
  }

  // Different styles/sizes
  &--inline-small {
    margin: 0 auto;
    height: 15px;
    width: 15px;
    top: 2px;
  }
}

@keyframes spinner-bounce {
  0%,
  100% {
    -webkit-transform: scale(0);
    transform: scale(0);
  }

  50% {
    -webkit-transform: scale(1);
    transform: scale(1);
  }
}
