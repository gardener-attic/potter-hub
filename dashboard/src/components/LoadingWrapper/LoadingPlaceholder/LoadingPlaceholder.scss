.LoadingPlaceholder {
  height: 1em;
  background-color: #f1f1f1;
  position: relative;
  &:after {
    content: "";
    position: absolute;
    width: 100%;
    height: 1em;
    background: linear-gradient(
      90deg,
      rgba(0, 0, 0, 0),
      rgba(255, 255, 255, 25%),
      rgba(0, 0, 0, 0)
    );
    transform: translateX(-30%);
    animation: loading 1.5s infinite;
  }
}

@keyframes loading {
  100% {
    transform: translateX(100%);
  }
}
