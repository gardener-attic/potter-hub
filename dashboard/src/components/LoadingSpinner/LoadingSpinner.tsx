import * as React from "react";

import "./LoadingSpinner.css";

export interface ISpinnerProps {
  size?: string;
}

const LoaderSpinner: React.SFC<ISpinnerProps> = props => {
  // Based on http://tobiasahlin.com/spinkit/
  return (
    <div className={`spinner ${props.size && `spinner--${props.size}`}`}>
      <div className="spinner__bounce1" />
      <div className="spinner__bounce2" />
    </div>
  );
};

export default LoaderSpinner;
