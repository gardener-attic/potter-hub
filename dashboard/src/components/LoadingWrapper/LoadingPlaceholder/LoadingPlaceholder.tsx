import * as React from "react";

import "./LoadingPlaceholder.css";

// Renders a placeholder loader style
// Based on https://kyusuf.com/post/fake-it-til-you-make-it-css/
const LoadingPlaceholder: React.SFC = props => {
  return <div className="LoadingPlaceholder" />;
};

export default LoadingPlaceholder;
