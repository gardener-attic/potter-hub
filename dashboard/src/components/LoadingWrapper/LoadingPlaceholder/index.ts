import LoadingPlaceholder from "./LoadingPlaceholder";

export default LoadingPlaceholder;
