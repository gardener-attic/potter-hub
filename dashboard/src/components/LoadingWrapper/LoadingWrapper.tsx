import * as React from "react";

import LoadingSpinner from "../LoadingSpinner";
import LoadingPlaceholder from "./LoadingPlaceholder";

export enum LoaderType {
  Spinner,
  Placeholder,
}

export interface ILoadingWrapperProps {
  type?: LoaderType;
  loaded?: boolean;
  size?: string;
}

// TODO(miguel) Move these kind of components to stateless compontents once we upgrade ts definitions
// Currently, I am having issues transforming it via const LoadingWrapper: React.SFC<ILoadingWrapperProps>
class LoadingWrapper extends React.Component<ILoadingWrapperProps> {
  public static defaultProps: ILoadingWrapperProps = {
    type: LoaderType.Spinner,
    loaded: false,
  };

  public render() {
    return this.props.loaded ? this.props.children : this.renderLoader();
  }

  private renderLoader() {
    switch (this.props.type) {
      case LoaderType.Spinner:
        return <LoadingSpinner size={this.props.size} />;
      case LoaderType.Placeholder:
        return <LoadingPlaceholder />;
    }
    return;
  }
}

export default LoadingWrapper;
