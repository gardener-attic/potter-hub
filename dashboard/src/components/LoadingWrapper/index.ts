import LoadingWrapper, { ILoadingWrapperProps as loadingWrapperProps, LoaderType } from "./LoadingWrapper";
export { LoaderType };
export type ILoadingWrapperProps = loadingWrapperProps;
export default LoadingWrapper;
