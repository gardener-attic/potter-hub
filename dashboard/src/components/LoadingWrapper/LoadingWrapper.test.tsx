import { shallow } from "enzyme";
import * as React from "react";
import LoadingWrapper, { LoaderType } from ".";
import LoadingSpinner from "../LoadingSpinner";
import LoadingPlaceholder from "./LoadingPlaceholder";

let props = {} as any;

const ChildrenComponent = () => <div>Hello dad!</div>;

const renderComponent = (p: any) => {
  return shallow(
    <LoadingWrapper {...p}>
      <ChildrenComponent />
    </LoadingWrapper>,
  );
};

describe("when loaded is false", () => {
  beforeEach(() => {
    props = {
      loaded: false,
    };
  });

  it("matches the snapshot", () => {
    const wrapper = renderComponent(props);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a spinner", () => {
    const wrapper = renderComponent(props);
    expect(wrapper.find(LoadingSpinner)).toExist();
  });

  it("does not render any children", () => {
    const wrapper = renderComponent(props);
    expect(wrapper.find(ChildrenComponent)).not.toExist();
  });
});

describe("when loaded is true", () => {
  beforeEach(() => {
    props = {
      loaded: true,
    };
  });

  it("matches the snapshot", () => {
    const wrapper = renderComponent(props);
    expect(wrapper).toMatchSnapshot();
  });

  it("does not renders a spinner", () => {
    const wrapper = renderComponent(props);
    expect(wrapper.find(LoadingSpinner)).not.toExist();
  });

  it("renders it wrapped component", () => {
    const wrapper = renderComponent(props);
    expect(wrapper.find(ChildrenComponent)).toExist();
  });
});

describe("loader types", () => {
  it("renders the LoadingPlaceholder type when specified", () => {
    const wrapper = renderComponent({ type: LoaderType.Placeholder });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(LoadingPlaceholder)).toExist();
  });
});
