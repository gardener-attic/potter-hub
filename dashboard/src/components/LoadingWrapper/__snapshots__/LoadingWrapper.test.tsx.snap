// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader types renders the LoadingPlaceholder type when specified 1`] = `<LoadingPlaceholder />`;

exports[`when loaded is false matches the snapshot 1`] = `<LoaderSpinner />`;

exports[`when loaded is true matches the snapshot 1`] = `<ChildrenComponent />`;
